base/asyncmap.jl

```diff
@@ -62,24 +72,34 @@ function done(itr::AsyncCollector, state::AsyncCollectorState)
 end
 
 function next(itr::AsyncCollector, state::AsyncCollectorState)
-    # Wait if the maximum number of concurrent tasks are already running
+    # Wait if the maximum number of concurrent tasks are already running.
     while isbusy(itr, state)
         wait(state)
     end
 
-    # Get index and mapped function arguments from enumeration iterator
+    # Get index and mapped function arguments from enumeration iterator.
     (i, args), state.enum_state = next(itr.enumerator, state.enum_state)
 
-    # Execute function call and save result asynchronously
+    # Execute function call and save result asynchronously.
     @async begin
-        itr.results[i] = itr.f(args...)
-        state.active_count -= 1
+        try
+            itr.results[i] = itr.f(args...)
+        catch e
+            state.in_error = true
+            notify(state.task_done, e; error=true)
```

Comment:
As I understand it this `notify` is not necessary for correctness does not improve performance to any significant degree (because without the `notify` the re-thrown `e` will be caught in `done` only a few lines after where the `wait` would have caught the `notify`).

I believe that getting error handling right in asynchronous code is a difficult thing, and that simplicity and ease of understanding the code is critical to getting it right. I am uncomfortable with this code because I think it makes the error handling appear to be more complicated than it really is. I think it is easier to follow what is going on if there is only one path for the error to be passed from the async task to the main iterator.

I think that the addition of the `in_error` flag to stop iteration early is a great improvement.
However I'd prefer it if the `notify` with `error=true` was removed and the existing `notify` was moved into the `finally` block.  Like this...
```julia
        try
            itr.results[i] = itr.f(args...)
        catch e
            # The in_error flag causes done() to end the iteration early and call sync_end().
            # sync_end() then re-throws "e" in the main task.
            state.in_error = true
            rethrow(e)
        finally
            state.active_count -= 1
            notify(state.task_done, nothing)
        end
```